### PR TITLE
chore: Update core storage to support React Native

### PR DESCRIPTION
**Pull request by cshfang**
#### Description of changes

This PR refactors core storage to support React Native

#### Description of how you validated changes
`yarn test`. Smoke tested in sample app

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/core/src/storage/CookieStorage.native.ts [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+import { KeyValueStorage } from './KeyValueStorage';
+
+/**
+ * @internal
+ */
+export class CookieStorage extends KeyValueStorage {
+	constructor() {
+		super(new InMemoryStorage());
+	}
+}
```
Comment (elorzafe) [inline packages/core/src/storage/CookieStorage.native.ts:14]: I am not sure this is appropiate, probably this should throw (no cookies on react-native)

File: packages/core/src/storage/utils.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+const TEST_KEY = 'aws.amplify.test-ls';
+
+/**
+ * @internal
+ * @returns Either a reference to window.localStorage or an in-memory storage as fallback
+ */
+export const getDefaultStorageWithFallback = (): Storage => {
+	try {
+		window.localStorage.setItem(TEST_KEY, '');
+		window.localStorage.removeItem(TEST_KEY);
```
Comment (elorzafe) [inline packages/core/src/storage/utils.ts:15]: Why do you need to test `localStorage`? Would not be enough just to check if the function exists?

File: packages/core/src/storage/utils.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+/**
+ * @internal
+ * @returns Either a reference to window.sessionStorage or an in-memory storage as fallback
+ */
+export const getSessionStorageWithFallback = (): Storage => {
+	try {
+		window.sessionStorage.setItem(TEST_KEY, '');
+		window.sessionStorage.removeItem(TEST_KEY);
+		return window.sessionStorage;
```
Comment (elorzafe) [inline packages/core/src/storage/utils.ts:30]: Same as above

**Review by elorzafe — CHANGES_REQUESTED** [review]
Overall looks very good, just one concern for how is tested if `localStorage` or `sessionStorage` is there

File: packages/core/src/storage/utils.ts (hunk 2 above)
Comment (cshfang) [reply packages/core/src/storage/utils.ts:15]: This was the way we've always done it and is a popular method for localStorage detection

**Review by AllanZhengYP — DISMISSED** [review]

**Review by elorzafe — APPROVED** [review]
Looks good!

Thanks @cshfang 🎖️

**Review by elorzafe — DISMISSED** [review]

**Review by AllanZhengYP — DISMISSED** [review]

File: packages/core/src/storage/DefaultStorage.ts [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+
+import { KeyValueStorage } from './KeyValueStorage';
+import { getDefaultStorageWithFallback } from './utils';
+
+/**
+ * @internal
+ */
+export class DefaultStorage extends KeyValueStorage {
+	constructor() {
+		super(getDefaultStorageWithFallback());
```
[the remark this answers is not in the document]
Comment (cshfang) [reply packages/core/src/storage/DefaultStorage.ts:12]: This is the original behavior as well

**Review by elorzafe — APPROVED** [review]

**Review by AllanZhengYP — APPROVED** [review]

**Pull request merged**